compiler/optimizer/OMRRecognizedCallTransformer.cpp

```diff
@@ -0,0 +1,58 @@
+/*******************************************************************************
+* Copyright (c) 2017, 2017 IBM Corp. and others
+*
+* This program and the accompanying materials are made available under
+* the terms of the Eclipse Public License 2.0 which accompanies this
+* distribution and is available at http://eclipse.org/legal/epl-2.0
+* or the Apache License, Version 2.0 which accompanies this distribution
+* and is available at https://www.apache.org/licenses/LICENSE-2.0.
+*
+* This Source Code may also be made available under the following Secondary
+* Licenses when the conditions for such availability set forth in the
+* Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+* version 2 with the GNU Classpath Exception [1] and GNU General Public
+* License, version 2 with the OpenJDK Assembly Exception [2].
+*
+* [1] https://www.gnu.org/software/classpath/license.html
+* [2] http://openjdk.java.net/legal/assembly-exception.html
+*
+* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+*******************************************************************************/
+
+#include "optimizer/RecognizedCallTransformer.hpp"
+
+#include "il/Node.hpp"
+#include "il/Node_inlines.hpp"
+#include "il/TreeTop.hpp"
+#include "il/TreeTop_inlines.hpp"
+#include "codegen/CodeGenerator.hpp" // for CodeGenerator
+
+TR::Optimization* OMR::RecognizedCallTransformer::create(TR::OptimizationManager *manager)
+   {
+   return new (manager->allocator()) TR::RecognizedCallTransformer(manager);
+   }
+
+int32_t OMR::RecognizedCallTransformer::perform()
+   {
+   for (auto treetop = comp()->getMethodSymbol()->getFirstTreeTop(); treetop != NULL; treetop = treetop->getNextTreeTop())
+      {
+      if (treetop->getNode()->getNumChildren() > 0)

```

Comment:
Should this somehow be checking for and only doing the transformation at the first occurrence of a commoned call node ? I am thinking of a case when more than one tree (regardless of the opcode at the top node, e.g. treetop or store) can have a particular call node as the first child and the first reference to it is the one where the call node will actually get evaluated. While it's possible that the project specific "transform" handles this somehow, should this be considered in the "tree walking" pass that is being considered at this (OMR) layer ?

Reply:
My apology for the late reply. This OMR layer code will call downstream project's "transform" for each occurrence of `call` node it sees; however, the project's "transform" may choose to transform or not. For example, should OpenJ9 uses it to transform `Unsafe`, OpenJ9's "transform" would probably only transform the `call` nodes marked in its fast-path. OMR does not have to and should not differentiate different occurrences of `call` node.

Reply:
Okay that sounds fine in theory but in practice I feel most downstream projects will end up duplicating the logic to avoid visiting the same call node multiple times. The call really "happens" at the first occurrence of the call node and so I'd actually argue the latter occurrences are not really calling anything and so no transformation should be trying to transform the later occurrences. Nor should it be necessary if the transformation of the call node happened correctly at the first occurrence.

Reply:
I see your point; however, by not limiting the transformation to the first occurrence, downstream projects can be more flexible. They can choose to perform different transformations in different occurrences. I feel it is improper to limit this flexibility. In addition, downstream projects can implement their own logic to avoid duplicating logic, such as checking visit count before transformation.

Reply:
I have to say that on principle I would agree with @vijaysun-omr in that uncommoning a node (ie transforming a non-first occurrence of a node) is a non-trivial operation and doesn't seem like it belongs with a utility pass like the RecognizedCallTransformer. If it is not ok to transform the first occurrence of a call, then some additional complex analysis will be needed to do the necessary uncommoning and then transform a subset. Unless there is a compelling use case in a downstream project for the additional functionality I would be more comfortable if the nodechecklist existed in OMR and the option only existed to transform the first instance of a call in place which would change the call evaluation which is the true intent of this utility.

Reply:
Thank you @vijaysun-omr and @andrewcraik for pointing out your concerns and ideas. I've updated the code so that now it only visits each call node once.